logging/writer.go

```diff
@@ -1,29 +1,70 @@
 package logging
 
-import "io"
+import (
+	"bytes"
+	"io"
+	"strings"
+	"sync"
+)
 
 // writer is an implementation of io.Writer that writes to a logger.
 type writer struct {
 	l Logger
 	f func(Logger, string)
+
+	mu  sync.Mutex
+	buf strings.Builder
 }
 
 // NewWriter returns an io.Writer that writes content to the given logger.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewWriter(l Logger) io.Writer {
```

Comment:
I suggest we change the implementation to an `io.WriteCloser` and have the `Close()` implementation flush any remaning buffered content to the logger, if there is any.